source/Variant.cpp

```diff
@@ -0,0 +1,265 @@
+#include "WeightedVariant.h"
+
+#include <algorithm>
+#include <iterator>
+
+using namespace std;
+
+
+
+// Construct and Load() at the same time.
+Variant::Variant(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Variant::Load(const DataNode &node)
+{
+	// If this variant is being loaded with a second token that is not a number,
+	// then it's being loaded from GameData, and therefore must have its
+	// name saved.
+	if(node.Size() >= 2 && !node.IsNumber(1))
+		name = node.Token(1);
+	
+	// If Load() has already been called once on this variant, any subsequent
+	// calls will replace the contents instead of adding to them.
+	bool reset = !variants.empty() || !ships.empty();
+	
+	for(const DataNode &child : node)
+	{
+		bool add = (child.Token(0) == "add");
+		bool remove = (child.Token(0) == "remove");
+		if((add || remove) && child.Size() == 1)
+		{	
+			child.PrintTrace("Skipping invalid \"" + child.Token(0) + "\" tag:");
+			continue;
+		}
+		bool variant = (child.Token(add || remove) == "variant");
+		
+		if(remove)
+		{
+			if(variant)
+			{
+				// If given a full definition of one of this fleet's variant members, remove the variant.
+				Variant toRemove(child);
+				auto VariantToRemove = [&](const WeightedVariant &v) noexcept -> bool
+				{
+					return v.Get() == toRemove;
+				};
+				
+				auto removeIt = remove_if(variants.begin(), variants.end(), VariantToRemove);
+				if(removeIt != variants.end())
+					variants.erase(removeIt, variants.end());
+				else
+					child.PrintTrace("Did not find matching variant for specified operation:");
+			}
+			else
+			{
+				// If given the name of a ship, remove all ships by that name from this variant.
+				string shipName = child.Token(1);
+				auto ShipToRemove = [&](const Ship *s) noexcept -> bool
+				{
+					return s->VariantName() == shipName;
+				};
+				
+				auto removeIt = remove_if(ships.begin(), ships.end(), ShipToRemove);
+				if(removeIt != ships.end())
+					ships.erase(removeIt, ships.end());
+				else
+					child.PrintTrace("Did not find matching ship for specified operation:");
+			}
+		}
+		else
+		{
+			// If this is a subsequent call of Load(), clear the variant
+			// if we aren't adding to it.
+			if(reset && !add)
+			{
+				reset = false;
+				variants.clear();
+				ships.clear();
+			}
+			
+			int n = 1;
+			int index = 1 + add;
+			if(variant)
+			{
+				string variantName;
+				if(child.Size() >= index + 1 && !child.IsNumber(index))
+				{
+					variantName = child.Token(index++);
+					if(variantName == name)
+					{
+						node.PrintTrace("A variant cannot reference itself:");
+						continue;
+					}
+				}
+				
+				if(child.Size() >= index + 1 && child.Value(index) >= 1.)
+					n = child.Value(index);
+				
+				// If this variant is named, then look for it in GameData.
+				// Otherwise this is a new variant definition only for this variant.
+				if(!variantName.empty())
+					variants.emplace_back(GameData::Variants().Get(variantName), n);
```

Comment:
It would be useful if we get a warning for the following:

```
fleet "Large Oathkeeper"
	variant
		variant "test"
			"Carrier (Mark IIIa)"
			"Carrier (Mark IIIb)"
                        # ^ the children are ignored since the variant named "test" is used
```